site/viamguides/src/pydantic-validation/pydantic-validation.md

```diff
@@ -0,0 +1,647 @@
+author: HipsterBrown
+id: pydantic-validation
+summary: Use Pydantic models to validate module configuration and environment variables
+categories: Developer, Validation, Modules
+environments: web
+status: Published 
+feedback link: https://github.com/viam-labs/viamcodelabs/issues
+tags: Modules, Developer, Validation, Intermediate
+
+# Module and environment variable validation with Pydantic
+<!-- ------------------------ -->
+## Overview 
+Duration: 1
+
+The [Viam Python SDK](https://python.viam.dev) is a great way to extend the platform with modules and automate machines with scripts. For each of these tasks, you may encounter the need to validate user-supplied configuration or settings from environment variables. While you can do this manually with simple [`type()`](https://docs.python.org/3/library/functions.html#type) and [`assert`](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement) checks, there are more robust, type-safe libraries for handling this logic such as [Pydantic](https://docs.pydantic.dev/latest/).
+
+![Pydantic logo text](./assets/pydantic-logo.svg)
+
+Pydantic is the most widely used data validation library for Python; used by HuggingFace, FastAPI, Django, and more.
+
+Building off the [Working with Python environment variables codelab](/guide/environment-variables/index.html), you'll learn about using Pydantic to ensure your Viam machines are configured correctly.
+
+### Prerequisites
+- Familiarity with Python
+
+### What You’ll Learn 
+- How to validate [modular resource](https://docs.viam.com/registry/#modular-resources) configurations
+- How to validate environment variable settings in a Python [process script](https://docs.viam.com/configure/processes/)
+
+### What You’ll Need 
+- A computer with MacOS, Windows, or Linux 
+- [Python3](https://www.python.org/downloads/) installed on your computer
+- [VS Code](https://code.visualstudio.com/download) installed, or another similar code editor of your choice.
+
+### What You’ll Build 
+- a [sensor component](https://docs.viam.com/components/sensor/) module with type-safe configuration
+- a Python script for automating a Viam machine that uses validated environment variables
+
+<!-- ------------------------ -->
+## Create a module project
+Duration: 5
+
+In this step, you'll build upon the How-to Guide for [creating a sensor module with Python](https://docs.viam.com/how-tos/sensor-module/) to provide robust, type-safe validation to this configuration logic.
+
+### Set up the Python sensor module project
+
+These instructions will not include the explanations of each file in the module project, see the linked How-to Guide for more of those details. If you use the [module generator](https://docs.viam.com/how-tos/sensor-module/#generate-boilerplate-module-code), you can skip to the 5th instruction in this section.
+
+1. Create the project directory through the command line or using the code editor of your choice:
```

Comment:
Thanks for calling out the ordered list tip when using only `1. `, I totally forgot about it. Definitely useful for consistency and updating later. 

> Also I prefixed terminal commands with $.
I forgo that prefix because it makes copying the command slightly annoying since you'd need to be careful not to include the $. 

> Semi-related, I'm curious what your thoughts are on compound commands using &&. Is it better to be explicit for readability, and list each step-by-step line-after-line? Or combine so you can copy and paste once?

This is a good question. I like one-liner commands but see the point of view where it could be confusing to someone who wants to understand exactly what's happening. There's also an argument for splitting into multiple lines so that folks reading the snippet don't need to horizontally scroll to see it all. It should still be possible to copy-paste if we don't include the $ prefix.